File: stock_web_app/engine/order_engine.py
from datetime import datetime
from typing import List, Dict
import threading
from engine.dsa.avl_tree import AVLTree
from engine.dsa.trie import Trie
from engine.dsa.depq import DEPQ
from engine.dsa.bplus_tree import BPlusTree

class OrderEngine:
    """
    Advanced Stock Order Matching Engine using DSA.
    - AVL Trees for price-sorted Order Book.
    - DEPQ for efficient Best Bid/Ask access.
    - Trie for stock symbol searching.
    - B+ Tree for trade history indexing.
    """
    
    def __init__(self, stock_symbols: List[str]):
        self.lock = threading.Lock()
        self.order_counter = 0
        self.trade_counter = 0
        
        # O(1) lookup by order_id
        self.orders_by_id = {}
        
        # AVL Trees for each stock (Price -> List of Orders)
        self.buy_trees = {symbol: AVLTree() for symbol in stock_symbols}
        self.sell_trees = {symbol: AVLTree() for symbol in stock_symbols}
        
        # DEPQ for Best Bid/Ask per stock
        self.depqs = {symbol: DEPQ() for symbol in stock_symbols}
        
        # Trie for stock symbol searching
        self.symbol_trie = Trie()
        for symbol in stock_symbols:
            self.symbol_trie.insert(symbol)
            
        # B+ Tree for Trade History indexing by timestamp
        self.trade_bplus_tree = BPlusTree(order=8)
        self.all_trades = [] # Keep list for quick reverse retrieval
    
    def place_order(self, symbol: str, order_type: str, price: float, quantity: int) -> Dict:
        """Place a new order using AVL Trees and DEPQ."""
        if quantity <= 0 or price <= 0:
            return {"success": False, "message": "Invalid price or quantity"}
        
        with self.lock:
            self.order_counter += 1
            order_id = f"ORD-{self.order_counter}"
            
            order = {
                "id": order_id,
                "symbol": symbol,
                "type": order_type.upper(),
                "price": float(price),
                "quantity": int(quantity),
                "original_quantity": int(quantity),
                "timestamp": datetime.now().isoformat()
            }
            
            # 1. Store in O(1) lookup
            self.orders_by_id[order_id] = order
            
            # 2. Store in AVL Tree (Price level)
            if order_type.upper() == "BUY":
                self.buy_trees[symbol].root = self.buy_trees[symbol].insert(self.buy_trees[symbol].root, price, order)
            else:
                self.sell_trees[symbol].root = self.sell_trees[symbol].insert(self.sell_trees[symbol].root, price, order)
            
            # 3. Update DEPQ
            self.depqs[symbol].insert(price, order_id)
            
        return {
            "success": True,
            "message": f"Order {order_id} placed in Order Book using AVL Tree.",
            "order_id": order_id
        }

    def process_market_orders(self, market_prices: Dict[str, float]) -> Dict:
        """Execute orders based on market price using AVL Tree traversal."""
        executed_trades = []
        
        with self.lock:
            for symbol, m_price in market_prices.items():
                # Process BUY orders (Market Price <= Order Price)
                # We need all nodes in BUY tree with price >= m_price
                self._match_buy_orders(symbol, m_price, executed_trades)
                
                # Process SELL orders (Market Price >= Order Price)
                # We need all nodes in SELL tree with price <= m_price
                self._match_sell_orders(symbol, m_price, executed_trades)
        
        return {
            "executed_count": len(executed_trades),
            "executed_trades": executed_trades
        }

    def _match_buy_orders(self, symbol, m_price, executed_trades):
        """Find and execute BUY orders where market_price <= limit_price."""
        tree = self.buy_trees[symbol]
        # We need to find all orders with price >= m_price
        # In a sorted tree, these are the nodes from m_price up to max
        # For simplicity, we can get inorder traversal and filter, 
        # or implement a more efficient range search in AVL.
        nodes = []
        tree.inorder(tree.root, nodes)
        
        for node in nodes:
            if node.price >= m_price:
                for order in list(node.orders):
                    self._execute_order(order, m_price, executed_trades)
                    node.orders.remove(order)
                    self.depqs[symbol].remove(order["id"])
                    del self.orders_by_id[order["id"]]
                
                if not node.orders:
                    tree.root = tree.delete(tree.root, node.price)

    def _match_sell_orders(self, symbol, m_price, executed_trades):
        """Find and execute SELL orders where market_price >= limit_price."""
        tree = self.sell_trees[symbol]
        # We need to find all orders with price <= m_price
        nodes = []
        tree.inorder(tree.root, nodes)
        
        for node in nodes:
            if node.price <= m_price:
                for order in list(node.orders):
                    self._execute_order(order, m_price, executed_trades)
                    node.orders.remove(order)
                    self.depqs[symbol].remove(order["id"])
                    del self.orders_by_id[order["id"]]
                
                if not node.orders:
                    tree.root = tree.delete(tree.root, node.price)

    def _execute_order(self, order, m_price, executed_trades):
        self.trade_counter += 1
        trade = {
            "trade_id": f"TRD-{self.trade_counter}",
            "order_id": order["id"],
            "symbol": order["symbol"],
            "type": order["type"],
            "limit_price": order["price"],
            "exec_price": float(m_price),
            "order_price": order["price"], # compatibility
            "market_price": float(m_price), # compatibility
            "quantity": order["quantity"],
            "timestamp": datetime.now().isoformat()
        }
        self.all_trades.append(trade)
        # Index in B+ Tree by timestamp (string)
        self.trade_bplus_tree.insert(trade["timestamp"], trade)
        executed_trades.append(trade)

    def get_order_book(self) -> Dict:
        """Get order book by traversing AVL Trees."""
        buy_side = []
        sell_side = []
        
        with self.lock:
            for symbol in self.buy_trees:
                # Traverse BUY AVL Tree
                nodes = []
                self.buy_trees[symbol].inorder(self.buy_trees[symbol].root, nodes)
                for node in nodes:
                    for order in node.orders:
                        buy_side.append({
                            "symbol": order["symbol"],
                            "price": order["price"],
                            "quantity": order["quantity"],
                            "id": order["id"]
                        })
                
                # Traverse SELL AVL Tree
                nodes = []
                self.sell_trees[symbol].inorder(self.sell_trees[symbol].root, nodes)
                for node in nodes:
                    for order in node.orders:
                        sell_side.append({
                            "symbol": order["symbol"],
                            "price": order["price"],
                            "quantity": order["quantity"],
                            "id": order["id"]
                        })
        
        # Sort for display (AVL inorder is sorted)
        buy_side.sort(key=lambda x: x["price"], reverse=True)
        sell_side.sort(key=lambda x: x["price"])
        
        return {
            "buy_side": buy_side,
            "sell_side": sell_side
        }

    def search_symbols(self, prefix: str) -> List[str]:
        """Use Trie for symbol prefix search."""
        return self.symbol_trie.search(prefix.upper())

    def get_trade_history(self, start_time: str = None, end_time: str = None) -> List[Dict]:
        """Get trade history, using B+ Tree for range queries if times provided."""
        if start_time and end_time:
            return self.trade_bplus_tree.get_range(start_time, end_time)
        return sorted(self.all_trades, key=lambda x: x["timestamp"], reverse=True)

    def get_stats(self) -> Dict:
        """Get engine statistics using AVL/Dict data."""
        with self.lock:
            total_buy_orders = sum(1 for o in self.orders_by_id.values() if o["type"] == "BUY")
            total_sell_orders = sum(1 for o in self.orders_by_id.values() if o["type"] == "SELL")
            total_buy_qty = sum(o["quantity"] for o in self.orders_by_id.values() if o["type"] == "BUY")
            total_sell_qty = sum(o["quantity"] for o in self.orders_by_id.values() if o["type"] == "SELL")
            
            return {
                "total_buy_orders": total_buy_orders,
                "total_sell_orders": total_sell_orders,
                "total_buy_quantity": int(total_buy_qty),
                "total_sell_quantity": int(total_sell_qty),
                "trades_executed": len(self.all_trades),
                "total_volume": sum(t["quantity"] for t in self.all_trades)
            }

    def get_best_prices(self, symbol: str) -> Dict:
        """Access best prices using DEPQ in O(1) for top element."""
        return {
            "best_bid": self.depqs[symbol].get_max(),
            "best_ask": self.depqs[symbol].get_min()
        }

File: stock_web_app/engine_cpp/order_engine.cpp
#include "order_engine.h"
#include <chrono>
#include <iomanip>
#include <sstream>
#include <iostream>

OrderEngine::OrderEngine(std::vector<std::string> symbols) : order_counter(0), trade_counter(0), stock_symbols(symbols) {
    for (const std::string& symbol : symbols) {
        buy_trees[symbol] = new AVLTree();
        sell_trees[symbol] = new AVLTree();
        symbol_trie.insert(symbol);
    }
}

OrderEngine::~OrderEngine() {
    for (auto const& pair : buy_trees) delete pair.second;
    for (auto const& pair : sell_trees) delete pair.second;
    for (Trade* t : all_trades) delete t;
}

std::string OrderEngine::current_time_str() {
    auto now = std::chrono::system_clock::now();
    auto in_time_t = std::chrono::system_clock::to_time_t(now);
    std::stringstream ss;
    ss << std::put_time(std::localtime(&in_time_t), "%Y-%m-%dT%H:%M:%S");
    return ss.str();
}

std::string OrderEngine::placeOrder(std::string symbol, std::string type, double price, int quantity) {
    order_counter++;
    std::string order_id = "ORD-" + std::to_string(order_counter);

    Order order = {order_id, symbol, type, price, quantity, current_time_str()};

    if (type == "BUY") {
        buy_trees[symbol]->insert(price, order);
    } else {
        sell_trees[symbol]->insert(price, order);
    }

    return order_id;
}

std::vector<Trade*> OrderEngine::processMarketOrders(std::unordered_map<std::string, double> market_prices) {
    std::vector<Trade*> executed_trades;

    for (auto const& pair : market_prices) {
        std::string symbol = pair.first;
        double m_price = pair.second;

        // Update price history for indicators
        price_history[symbol].push_back(m_price);
        if (price_history[symbol].size() > MAX_WINDOW) {
            price_history[symbol].pop_front();
        }

        // Match BUY orders (Order Price >= Market Price)
        AVLTree* buy_tree = buy_trees[symbol];
        std::vector<AVLNode*> buy_nodes;
        buy_tree->inorder(buy_nodes);

        for (int i = buy_nodes.size() - 1; i >= 0; --i) {
            AVLNode* node = buy_nodes[i];
            if (node->price >= m_price) {
                for (const auto& order : node->orders) {
                    trade_counter++;
                    Trade* t = new Trade {
                        "TRD-" + std::to_string(trade_counter),
                        order.id, symbol, order.type, order.price, m_price, order.quantity, current_time_str()
                    };
                    executed_trades.push_back(t);
                    all_trades.push_back(t);
                    trade_index.insert(t->timestamp, t);
                }
                buy_tree->remove(node->price);
            }
        }

        // Match SELL orders (Order Price <= Market Price)
        AVLTree* sell_tree = sell_trees[symbol];
        std::vector<AVLNode*> sell_nodes;
        sell_tree->inorder(sell_nodes);

        for (AVLNode* node : sell_nodes) {
            if (node->price <= m_price) {
                for (const auto& order : node->orders) {
                    trade_counter++;
                    Trade* t = new Trade {
                        "TRD-" + std::to_string(trade_counter),
                        order.id, symbol, order.type, order.price, m_price, order.quantity, current_time_str()
                    };
                    executed_trades.push_back(t);
                    all_trades.push_back(t);
                    trade_index.insert(t->timestamp, t);
                }
                sell_tree->remove(node->price);
            }
        }
    }

    return executed_trades;
}

std::map<std::string, std::vector<std::vector<Order>>> OrderEngine::getOrderBook() {
    std::map<std::string, std::vector<std::vector<Order>>> book;
    std::vector<std::vector<Order>> buy_side;
    std::vector<std::vector<Order>> sell_side;

    for (const std::string& symbol : stock_symbols) {
        std::vector<AVLNode*> b_nodes;
        buy_trees[symbol]->inorder(b_nodes);
        for (auto* n : b_nodes) buy_side.push_back(n->orders);

        std::vector<AVLNode*> s_nodes;
        sell_trees[symbol]->inorder(s_nodes);
        for (auto* n : s_nodes) sell_side.push_back(n->orders);
    }
    
    book["buy"] = buy_side;
    book["sell"] = sell_side;
    return book;
}

std::vector<Trade*> OrderEngine::getTrades(std::string start, std::string end) {
    if (!start.empty() && !end.empty()) {
        return trade_index.rangeSearch(start, end);
    }
    return all_trades;
}

double OrderEngine::calculateSMA(std::string symbol, int period) {
    if (price_history[symbol].size() < (size_t)period) return -1.0;
    double sum = 0;
    auto it = price_history[symbol].rbegin();
    for (int i = 0; i < period; ++i, ++it) {
        sum += *it;
    }
    return sum / period;
}

double OrderEngine::calculateRSI(std::string symbol, int period) {
    if (price_history[symbol].size() < (size_t)period + 1) return -1.0;
    
    double gain = 0, loss = 0;
    auto it = price_history[symbol].rbegin();
    for (int i = 0; i < period; ++i) {
        double diff = *it - *(it + 1);
        if (diff > 0) gain += diff;
        else loss -= diff;
        ++it;
    }
    
    if (loss == 0) return 100.0;
    double rs = (gain / period) / (loss / period);
    return 100.0 - (100.0 / (1.0 + rs));
}

std::unordered_map<std::string, long long> OrderEngine::getStats() {
    long long total_buy_orders = 0;
    long long total_sell_orders = 0;
    long long total_buy_qty = 0;
    long long total_sell_qty = 0;

    for (const std::string& symbol : stock_symbols) {
        std::vector<AVLNode*> b_nodes;
        buy_trees[symbol]->inorder(b_nodes);
        for (auto* n : b_nodes) {
            total_buy_orders += n->orders.size();
            for (auto& o : n->orders) total_buy_qty += o.quantity;
        }

        std::vector<AVLNode*> s_nodes;
        sell_trees[symbol]->inorder(s_nodes);
        for (auto* n : s_nodes) {
            total_sell_orders += n->orders.size();
            for (auto& o : n->orders) total_sell_qty += o.quantity;
        }
    }

    std::unordered_map<std::string, long long> stats;
    stats["total_buy_orders"] = total_buy_orders;
    stats["total_sell_orders"] = total_sell_orders;
    stats["total_buy_quantity"] = total_buy_qty;
    stats["total_sell_quantity"] = total_sell_qty;
    stats["trades_executed"] = (long long)all_trades.size();
    
    long long volume = 0;
    for (size_t i = 0; i < all_trades.size(); ++i) volume += all_trades[i]->quantity;
    stats["total_volume"] = volume;

    return stats;
}

std::vector<std::string> OrderEngine::searchSymbols(std::string prefix) {
    return symbol_trie.search(prefix);
}
